04-materials/05-developing-with-ai.md

```diff
@@ -1,14 +1,1290 @@
+
+AI coding assistants are powered by **Large Language Models ({term}`LLMs <LLM>`)** — neural networks trained on vast amounts of text and code. These models can:
+- Understand context from your codebase
+- Generate code snippets based on natural language descriptions
+- Explain existing code and suggest improvements
+- Debug errors by analyzing stack traces and code patterns
+
+### ☁️ Where {term}`LLMs <LLM>` live: deployment models
+
+**Frontier Models (Cloud-Hosted):**
+- **Deployment:** Run on massive server infrastructure by model providers
+- **Access:** Pay-per-{term}`token <LLM token>` via API keys
+- **Examples:**
+  - **Claude Sonnet 4.5** (Anthropic): Best coding model, \$3/\$15 per 1M tokens
+  - **GPT-5** (OpenAI): Best overall reasoning, \$1.25/\$10 per 1M tokens
+  - **Gemini 2.5 Pro** (Google): Best for speed/context (1M tokens), multimodal, \$1.25/\$10 per 1M tokens
+  - **Claude 4 Opus** (Anthropic): Best for long-horizon coding (30+ hour tasks), \$15/\$75 per 1M tokens
+- **Pros:** State-of-the-art capabilities, specialized for different tasks (coding vs reasoning vs speed), no local compute needed
+- **Cons:** Requires internet connection, ongoing costs, data leaves your machine
+
+**Mid-tier and efficient models (cloud or local):**
+- **Deployment:** Can run on cloud APIs or self-hosted on consumer hardware
+- **Examples:**
+  - **Claude Haiku**
+  - **Qwen3-30B-A3B** (approaching GPT-4o performance),
+  - **Mistral Small 3.2**
+  - **Llama 3.3-70B**
+- **Pros:** Lower cost or free (if self-hosted), faster responses, good balance of capability and efficiency
+- **Cons:** Less capable than frontier models, self-hosting requires GPU resources (typically 16GB+ {term}`VRAM`)
+
+
+**Open-source & open-weight models (2025 state-of-the-art):**
+- **Licenses:** Vary from fully open (Apache 2.0, MIT) to restricted commercial use
+- **Deployment:** Can be self-hosted using tools like [Ollama](https://ollama.com/), [LM Studio](https://lmstudio.ai/), or [vLLM](https://github.com/vllm-project/vllm)
+- **Examples:**
+  - **Qwen3-235B-A22B** (Apache 2.0 license): 235B params with 22B active, 262K context, exceptional reasoning
+  - **GLM-4.5** (Open License): Strong coding and agentic abilities, runs on consumer hardware
+  - **GLM-4.5 Air**: Optimized for 48GB RAM laptops when quantized
+  - **Qwen3-Coder**: Specialized for code generation tasks
+  - **DeepSeek-R1**: 671B params (37B active), MIT license, advanced reasoning (86.7% on AIME)
+  - **OpenAI GPT-OSS-120B/20B** (Apache 2.0): Near o4-mini performance, consumer-friendly
+- **Pros:** Full control, no API costs, data stays local, latest open models often outperform closed frontier ones
+- **Cons:** Requires technical setup and adequate hardware (16GB+ VRAM for smaller models, 48GB+ for larger ones)
+
+:::{dropdown} 💡 Self-Hosting {term}`LLMs <LLM>` (Optional)
+If you want to run models locally (for privacy or cost savings), tools like [Ollama](https://ollama.com/) make it easy:
```

Comment:
Maybe more detail on privacy is outside the scope of the tutorial, but it might be worth providing a link to some resources with more explanations about privacy concerns? Or just a term/definition of what we mean when we say "for privacy"?